Label Widget: Improve word wrapping and adding an word_wrap_mode

vinicius-toshiyuki opened the pull request:
The word_wrap_mode configuration can be set to
GUI::WordWrapMode::Word (default) or GUI::WordWrapMode::Anywhere.
GUI::WordWrapMode::Word will wrap text on spaces, without breaking
words, and long words that won't fit in a full line will be elided.
GUI::WordWrapMode::Anywhere will wrap text using all the available
space in a line, breaking mid-word if necessary.

kleinesfilmroellchen commented:
Please remove the merge commit and squash your other commits as per the code submission policy. I'll take a look at it after that :^)

vinicius-toshiyuki commented:
> Please remove the merge commit and squash your other commits as per the code submission policy. I'll take a look at it after that :^)

Removed the merge, squashed commits and passed in the lint_commits CI job

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -135,18 +139,34 @@ void Label::wrap_text()
 {
     Vector<String> words;
     Optional<size_t> start;
+    bool word_is_space = false;
+    auto include_word = [this, &start, &words](size_t i) {
+        if (start.has_value())
+            words.append(m_text.substring(start.value(), i - start.value()));
+        start.clear();
+    };
     for (size_t i = 0; i < m_text.length(); i++) {
         switch (m_text[i]) {
-        case '\n':
         case '\r':
+        case '\n': {
```
Do not unnecessarily reorganize code if it serves no purpose.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
     m_lines = lines;
 }
 
+size_t Label::next_substring_size(String word, size_t start)
```
If avoidable here, let's not pass a string by value in such a helper function where the string is only of temporary use anyways.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
+size_t Label::next_substring_size(String word, size_t start)
+{
+    String sub_word = word.substring(start);
+    size_t lower_idx = 0, upper_idx = sub_word.length();
+    size_t idx = (upper_idx - lower_idx) / 2;
```
Please avoid the [`idx` shortening](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#names), we're not on a 80-character limit :^)

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
+                continue;
+            }
+            upper_idx = idx;
+            idx -= (idx - lower_idx) / 2;
```
For clarity, I'd recommend using `upper_idx` here because this is just the same logic as the initialization on line 225.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
+            if (lower_idx == idx)
+                idx++;
+            lower_idx = idx;
+            idx += (upper_idx - idx) / 2;
```
Same as above, let's use `lower_idx` for clarity.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.cpp:
```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
+            upper_idx = idx;
+            idx -= (idx - lower_idx) / 2;
+        } else if (sub_word_width < rect.width()) {
+            if (lower_idx == idx)
+                idx++;
```
Is there a particular reason for not `continue`ing here?

vinicius-toshiyuki replied:
Yeah, with a `continue` there may be an infinite loop that goes like that:
- `lower_idx == idx` then `idx++` and `continue`
- enters the `sub_word_width > rect.width()` `if` and `upper_idx == idx`
- `idx--`, `continue` and repeat

kleinesfilmroellchen replied:
Oh, okay, fine then :^)

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.h:
```diff
@@ -36,6 +37,9 @@ class Label : public Frame {
     bool is_word_wrap() const { return m_word_wrap; }
     void set_word_wrap(bool);
 
+    GUI::WordWrapMode word_wrap_mode() const { return m_word_wrap_mode; }
```
This can be "full const" if I see that correctly.
```suggestion
    GUI::WordWrapMode const word_wrap_mode() const { return m_word_wrap_mode; }
```

vinicius-toshiyuki replied:
I get some compile errors with this change, is it really necessary? `text_alignment()` doesn't have it and both getters have the same logic..

kleinesfilmroellchen replied:
Nevermind I guess

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/Label.h:
```diff
@@ -47,13 +51,15 @@ class Label : public Frame {
 private:
     void size_to_fit();
     void wrap_text();
+    size_t next_substring_size(String, size_t);
```
See comment for not passing a string by value.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/WordWrapMode.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace GUI {
+
+#define GUI_ENUMERATE_WORD_WRAP_MODES(M) \
+    M(Word)                              \
+    M(Anywhere)
```
I know we do this in a couple of places, but is it *really* necessary for a two-member enum to use convenience enumeration macros? I'd like to hear what the maintainers say about this.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/WordWrapMode.h:
```diff
@@ -0,0 +1,44 @@
+#undef __ENUMERATE
+};
+
+inline Optional<WordWrapMode> word_wrap_mode_from_string(const StringView& string)
```
Simple `inline`s are often ignored by the compiler and therefore not useful for things we want to inline all the time. Let's use the `ALWAYS_INLINE` macro. Also, east const.
```suggestion
ALWAYS_INLINE Optional<WordWrapMode> word_wrap_mode_from_string(StringView const& string)
```

vinicius-toshiyuki replied:
I wrote this file based on <LibGfx/TextAlignment.h>, which also uses simple `inlines`, so maybe those should be changed too

kleinesfilmroellchen replied:
That's for another PR :yakstack:

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/WordWrapMode.h:
```diff
@@ -0,0 +1,44 @@
+    return {};
+}
+
+inline const char* to_string(WordWrapMode mode)
```
Same two remarks as above.
```suggestion
ALWAYS_INLINE char const* to_string(WordWrapMode mode)
```

kleinesfilmroellchen reviewed (changes requested):
I'm not confident I understand the logic behind `next_substring_size`, but I trust that it makes sense :^). Otherwise, this is my first-level review that definitely needs someone else to have a look as well.
